test/behavior/saturating_arithmetic.zig

```diff
@@ -0,0 +1,154 @@
+const std = @import("std");
+const builtin = @import("builtin");
+const mem = std.mem;
+const expectEqual = std.testing.expectEqual;
+const Vector = std.meta.Vector;
+const minInt = std.math.minInt;
+const maxInt = std.math.maxInt;
+
+test "@addWithSaturation" {
+    const S = struct {
+        fn doTheTest() !void {
+            const test_data = .{
+                // { a, b, expected a+b }
+                [_]i8{ -3, 10, 7 },
+                [_]i8{ -128, -128, -128 },
+                [_]i2{ 1, 1, 1 },
+                [_]i64{ maxInt(i64), 1, maxInt(i64) },
+                [_]i128{ maxInt(i128), -maxInt(i128), 0 },
+                [_]i128{ minInt(i128), maxInt(i128), -1 },
+                [_]i8{ 127, 127, 127 },
+                [_]u8{ 3, 10, 13 },
+                [_]u8{ 255, 255, 255 },
+                [_]u2{ 3, 2, 3 },
+                [_]u3{ 7, 1, 7 },
+                [_]u128{ maxInt(u128), 1, maxInt(u128) },
+            };
+
+            inline for (test_data) |array| {
```

Comment:
Since you're doing an `inline for` here, you're only testing the comptime code path and not testing the runtime code path. Make it a normal `for` loop to gain test coverage for the LLVM IR instructions being emitted.

Reply:
Oops! Really glad you caught this.  This uncovers a bug with the way @mulWithSaturation is calling the llvm intrinsic.  After reorganizing the tests, I'm now seeing a bunch of these, one for each of the types being used i believe.

```console
broken LLVM module found: 
Intrinsic name not mangled correctly for type arguments! Should be: llvm.smul.fix.sat.i64
i64 (i64, i64, i32)* @llvm.smul.fix.sat.i64.i64.p0i32
```

I'm not yet sure how to fix this :thinking: 

Here is the call site:
https://github.com/ziglang/zig/pull/9619/files#diff-bc5f44920e8a408b813211358b86bb1e11b8a6fd2ad74db7cd512c12264731acR491

EDIT:
I figured it out.  the solution was to get rid of the extra 2 type parameters at the call site.  This uncovers an issue with @mulWithSaturation :smirk: 

```zig
try testSaturatingOp(.mul, i8, .{ -128, -128, 127 });
```
produces
```console
... 
expected 127, found -128
...
```

I verified that the args are being passed correctly to llvm:
```zig
extern fn @"llvm.smul.fix.sat.i8"(i8, i8, i32) i8;
test "satmul" {
    const x = @"llvm.smul.fix.sat.i8"(-128, -128, 0);
    try std.testing.expectEqual(@as(i8, 127), x);
}
```
produces the same
```console
expected 127, found -128
```
Does this look like an llvm bug or am i missing something?

Reply:
We talked about this on discord.  Here are the steps to andrew suggested to move forward.
1. check if the bug is fixed in llvm 13 (I can check for you if you give me something to copy-paste)
2. file llvm bug report using only LLVM IR
3. in your PR, don't use the intrinsic for multiplication. instead, generate different code that accomplishes the same semantics (e.g. do the multiplication with more bits, then do some max/min stuff, then truncate)
4. make sure we have an open zig bug report to track the upstream llvm bug, so that when it's solved, we can update the code to use the intrinsic

Reply:
I created #9643 to track the upstream bug.  Its my first llvm bug report so I'm waiting for a new user registration email reply.  I have commented the mul test cases which trigger the upstream bug and will now work on 3. from above.

Reply:
I created the [upstream bug report](https://bugs.llvm.org/show_bug.cgi?id=51677).  I'm not yet sure how to proceed on 3 from above.  Codegen for the multiplication with extra bits seems simple enough, but not the max/min calculations and truncation.  Any ideas?  I'm hoping not to spend too much time on this temporary solution.

Reply:
max/min calculations are also described as widening in formal methods literature and include for 2 numbers bigInt and i32: -bigInt*i32, bigInt*-i32, bigInt*i32, bigInt*-i32 and the respective choosing of those (by the sign) to compare results against the min/max.

For u32 this cuts the necessary checks in half.

What do you mean with truncation?

Reply:
I'm looking for ideas about how to do temporary codegen for @mulWithSaturation in place of the llvm.smul.fix.sat intrinsic until the upstream bug is fixed.  I don't know of any issues with the unsigned version (llvm.smul.fix.sat) so this would only have to work for signed ints.

Reply:
It looks like the commented out tests pass in release-fast and release-safe modes (but not debug and release-small).  Perhaps we can leave things as is and temporarily skip the failing test cases in debug and release-small modes?  

If so, i could add a note in the langref that @mulWithSaturation has these limitations with a link to #9643.  Also, a TODO in the tests to remove the mode check once #9643 is resolved.  This might be like a decent alternative to implementing a temporary codegen solution.

Reply:
We discussed on discord. plan is skipping the failing tests (in all modes) with a comment linking to the zig issue tracking the upstream issue and mentioning limitations in the langref with link to #9643